distribution/lib/Standard/Base/0.0.0-dev/src/System/File/Generic/Writable_File.enso

```diff
@@ -27,44 +27,66 @@ from project.Data.Boolean import Boolean, False, True
 
     ## PRIVATE
     write self (existing_file_behavior : Existing_File_Behavior) (action : Output_Stream -> Any) -> Any =
-        self.write_strategy.write self.file existing_file_behavior action
+        self.write_strategy.write self.file existing_file_behavior action |> self.replace_resulting_file
 
     ## PRIVATE
     write_handling_dry_run self (existing_file_behavior : Existing_File_Behavior) (action : Any -> Output_Stream -> Any) -> Any =
-        self.write_strategy.write_handling_dry_run self.file existing_file_behavior action
+        self.write_strategy.write_handling_dry_run self.file existing_file_behavior action |> self.replace_resulting_file
 
     ## PRIVATE
-    is_local self -> Boolean = self.file.is_a File
+    is_local self -> Boolean =
+        ## Note: if we ever get data links (or other proxies) pointing to local files; such a proxy with a local file
+          target should probably not be considered local, because that may come with too many assumptions.
+        self.file.is_a File
 
     ## PRIVATE
        This helper method may be used if the logic for writing the file requires a local filesystem file
        (e.g. using some external libraries that prefer to work with a Java File).
        It downloads a remote file to a local temporary location, runs the action
        and then uploads the resulting file back to the remote backend.
     write_requiring_local_file self (existing_file_behavior : Existing_File_Behavior) (action : File -> Any) -> Any =
-        self.write_strategy.write_with_local_file self.file existing_file_behavior action
+        self.write_strategy.write_with_local_file self.file existing_file_behavior action |> self.replace_resulting_file
 
     ## PRIVATE
        Writes a local file to this `Writable_File` destination.
        This is used by `File.copy_to` and `File.move_to` to possibly implement
        the upload more efficiently (avoiding duplicated temporary files).
     copy_from_local self (source : File) (replace_existing : Boolean) =
-        self.write_strategy.copy_from_local source self.file replace_existing
+        self.write_strategy.copy_from_local source self.file replace_existing |> self.replace_resulting_file
 
     ## PRIVATE
     with_output_stream self (open_options : Vector) action =
         self.file.with_output_stream open_options action
 
+    ## PRIVATE
+       Checks the returned file, and if `self.file` is to be returned it is
+       replaced with `self.file_for_return`.
+
+       This ensures correct behaviour in case of proxies (e.g. data links). If
+       the original file is supposed to be returned, it is replaced with the
+       proxy. If another file is returned (e.g. due to dry-run), then it is
+       returned without changes.
+    replace_resulting_file self file_to_be_returned =
+        if file_to_be_returned == self.file then self.file_for_return else file_to_be_returned
```

Comment:
I wasn't 100% sure about this as it felt a little bit like a 'hack'. 

It felt like the `action` of the `write` operation should know what right value to return, and it felt like it is not good to narrow down return values to just the written file - in some cases the write action could want to return something else entirely.

However, after consideration I think it's probably OK.

The thing is - the write `action` is low level - it is encapsulating _how_ to write data to some destination. At this point it does not know if it's writing through a datalink or not. The best it can do when `write_handling_dry_run` is to return the `effective_file`, because the original file target might have been replaced with a dry-run file. We cannot however put the datalink proxy file as `effective_file` argument, because as mentioned - this is the low level operation and it expects this to be the actual file written to, not a proxy. So `effective_file` must stay to be the actual written file. We _could_ add one more argument - `original_file` (which would be the datalink) and let the `action` decide to return `effective_file` OR `original_file`. But that will complicate the actions and overall API and it feels unnecessary.

For now I decided to introduce a convention that the write operations return the actually written-to file (this was de facto already true). Thus, the `Writable_File` wrappers are able to replace this returned file with the proxy, and they are also able to _not_ replace it if a dry-run file was returned (which will be `!=` from the original _target_).

I think this makes it simpler and ensures that the implementations of `File_Write_Strategy` do not need to care that much about replacing the file - as it's handled in one place here.

Moreover, if we ever want to break the _convention_ and return something else than the file - the `replace_resulting_file` will simply pass such values forward unchanged, so there shouldn't be an issue with that either.